dpar/src/lookup.rs

```diff
@@ -0,0 +1,380 @@
+    /// are looked up.
+    ///
+    /// ```
+    /// use dpar::lookup::{FreezableLookup, LookupValue};
+    ///
+    /// let lookup = FreezableLookup::default().freeze();
+    /// assert_eq!(lookup.lookup("hello"), 0);
+    /// let thawed = lookup.thaw();
+    /// assert_eq!(thawed.lookup("hello"), 1);
+    /// ```
+    pub fn thaw(self) -> Self {
+        use self::FreezableLookupInner::*;
+
+        let inner = match self.0 {
+            Frozen(numberer) => Fresh(numberer.into()),
+            fresh => fresh,
+        };
+
+        FreezableLookup(inner)
+    }
+}
+
+impl<T> Default for FreezableLookup<T>
+where
+    T: Eq + Hash,
+{
+    fn default() -> Self {
+        FreezableLookup(FreezableLookupInner::Fresh(RefCell::new(Numberer::new(1))))
+    }
+}
+
+impl<O> LookupLen for FreezableLookup<O>
+where
+    O: Eq + Hash,
+{
+    fn len(&self) -> usize {
+        use self::FreezableLookupInner::*;
+
+        match &self.0 {
+            Fresh(cell) => cell.borrow().len() + 1,
+            Frozen(numberer) => numberer.len() + 1,
+        }
+    }
+}
+
+impl<B, O> LookupValue<B> for FreezableLookup<O>
+where
+    B: ?Sized + Eq + Hash + ToOwned<Owned = O>,
+    O: Borrow<B> + Clone + Eq + Hash,
+{
+    fn lookup(&self, t: &B) -> usize {
+        use self::FreezableLookupInner::*;
+
+        match &self.0 {
+            Fresh(cell) => cell.borrow_mut().add(t),
+            Frozen(numberer) => numberer.number(t).unwrap_or(0),
+        }
+    }
+
+    fn value(&self, id: usize) -> Option<Cow<B>> {
+        use self::FreezableLookupInner::*;
+
+        if id == 0 {
+            return None;
+        }
+
+        match &self.0 {
+            Fresh(cell) => cell.borrow().value(id).map(|v| Cow::Owned(v.to_owned())),
+            Frozen(numberer) => numberer.value(id).map(|v| Cow::Borrowed(v.borrow())),
+        }
+    }
+}
+
+impl<T> Serialize for FreezableLookup<T>
+where
+    T: Eq + Hash + Serialize,
+{
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        use self::FreezableLookupInner::*;
+
+        match &self.0 {
+            Fresh(refcell) => serializer.serialize_newtype_variant(
+                "FreezableLookup",
+                1,
+                "Frozen",
+                &*refcell.borrow(),
+            ),
+            Frozen(ref numberer) => {
+                serializer.serialize_newtype_variant("FreezableLookup", 1, "Frozen", numberer)
+            }
+        }
+    }
+}
+
+/// Numberer for categorical values, such as features or class labels.
+#[derive(Debug, Eq, PartialEq, Serialize, Deserialize)]
+struct Numberer<T>
+where
+    T: Eq + Hash,
+{
+    values: Vec<T>,
+    numbers: HashMap<T, usize>,
+    start_at: usize,
+}
+
+impl<T> Numberer<T>
+where
+    T: Eq + Hash,
+{
+    pub fn new(start_at: usize) -> Self {
+        Numberer {
+            values: Vec::new(),
+            numbers: HashMap::new(),
+            start_at: start_at,
```

Comment:
start_at assignment is unnecessary

Reply:
Indeed, copy and paster from code before this was supported in Rust.